extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/photos/GooglePhotosImporter.java

```diff
@@ -40,92 +42,126 @@
+      GooglePhotosInterface photosInterface,
       ImageStreamProvider imageStreamProvider) {
     this.credentialFactory = credentialFactory;
     this.jobStore = jobStore;
-    this.photosService = photosService;
+    this.jsonFactory = jsonFactory;
+    this.photosInterface = photosInterface;
     this.imageStreamProvider = imageStreamProvider;
   }
 
   @Override
   public ImportResult importItem(UUID jobId, TokensAndUrlAuthData authData,
-      PhotosContainerResource data) throws IOException, ServiceException {
+      PhotosContainerResource data) throws IOException {
+    // Uploads album metadata
     if (data.getAlbums() != null && data.getAlbums().size() > 0) {
-      logger.warn(
-          "Importing albums in Google Photos is not supported. "
-              + "Photos will be added to the default album.");
+      for (PhotoAlbum album : data.getAlbums()) {
+        importSingleAlbum(jobId, authData, album);
+      }
     }
 
+    // Uploads photos
     if (data.getPhotos() != null && data.getPhotos().size() > 0) {
       for (PhotoModel photo : data.getPhotos()) {
-        importSinglePhoto(authData, photo, jobId);
+        importSinglePhoto(jobId, authData, photo);
       }
     }
 
     return ImportResult.OK;
   }
 
   @VisibleForTesting
-  void importSinglePhoto(TokensAndUrlAuthData authData, PhotoModel inputPhoto, UUID jobId)
-      throws IOException, ServiceException {
-
-    // Set up photo
-    PhotoEntry outputPhoto = new PhotoEntry();
-    outputPhoto.setTitle(new PlainTextConstruct("copy of " + inputPhoto.getTitle()));
-    outputPhoto.setDescription(new PlainTextConstruct(inputPhoto.getDescription()));
-    outputPhoto.setClient(GoogleStaticObjects.APP_NAME);
-
-    String mediaType = inputPhoto.getMediaType();
-    if (mediaType == null) {
-      mediaType = "image/jpeg";
+  void importSingleAlbum(UUID jobId, TokensAndUrlAuthData authData, PhotoAlbum inputAlbum)
+      throws IOException {
+    // Set up album
+    GoogleAlbum googleAlbum = new GoogleAlbum();
+    googleAlbum.setTitle(COPY_PREFIX + inputAlbum.getName());
+
+    GoogleAlbum responseAlbum = getOrCreatePhotosInterface(authData).createAlbum(googleAlbum);
+    TempPhotosData tempPhotosData = jobStore
+        .findData(jobId, TEMP_PHOTOS_KEY, TempPhotosData.class);
+    if (tempPhotosData == null) {
+      tempPhotosData = new TempPhotosData(jobId);
+      jobStore.create(jobId, TEMP_PHOTOS_KEY, tempPhotosData);
     }
+    tempPhotosData.addAlbumId(inputAlbum.getId(), responseAlbum.getId());
+    jobStore.update(jobId, TEMP_PHOTOS_KEY, tempPhotosData);
+  }
 
-    MediaStreamSource streamSource;
+  @VisibleForTesting
+  void importSinglePhoto(UUID jobId, TokensAndUrlAuthData authData, PhotoModel inputPhoto)
+      throws IOException {
+    /*
+    TODO: resumable uploads https://developers.google.com/photos/library/guides/resumable-uploads
+    Resumable uploads would allow the upload of larger media that don't fit in memory.  To do this,
```

Comment:
Thanks for the comment. Are we erroring properly for files that are too large to fit in memory?

Reply:
I believe that outputStream.toByteArray() will error out in that case.